### ISSUE TITLE
Peer/PeerID issues.

### ISSUE THREAD

**Issue by cheatfate** [status-im/nim-libp2p#34]
In recently merged PR https://github.com/status-im/nim-libp2p/pull/32 there was made changes in peer.nim which i think needs to be reconsidered.

1. https://github.com/status-im/nim-libp2p/blob/master/libp2p/peer.nim#L21-L25
```nim
type
  PeerID* = object
    data*: seq[byte]
    privateKey*: Option[PrivateKey]
    publicKey: Option[PublicKey]
```
PeerID type now has optional fields of `privateKey` and `publicKey`

#### privateKey

Field `privateKey` field is currently used, and will be used only to represent and store local peer's private key. All other peers which will be stored in memory, database will have this field empty because we will not know private keys of remote nodes. Size of empty `Option[PrivateKey]` is 80 bytes.

#### publicKey

Now about `publicKey` field. PeerID itself is base58 encoded string of `multihash(sha256, "remote peer's public key")`, this is done for all public keys which size exceeds 42 bytes (e.g. RSA, ECNIST, which are the most used keys in current network), public keys from secp256k1 or ed25519 which is less then 42 bytes it will be stored as not *hashed* `multihash(identity, "remote peer's public key").

So in 50% of cases public key could not be recovered from PeerID, to check if remote PeerID has public key and to retrieve this public key there was present two procedures:

- `proc hasPublicKey*(pid: PeerID): bool` https://github.com/status-im/nim-libp2p/blob/master/libp2p/peer.nim#L98
- `proc extractPublicKey*(pid: PeerID, pubkey: var PublicKey): bool` https://github.com/status-im/nim-libp2p/blob/master/libp2p/peer.nim#L106

IMHO this two procedures are enough to retrieve public key from PeerID if its needed.

To store empty `Option[PublicKey]` you also need 80 bytes, initialized with RSA/ECNIST public keys it will takes much more.

So total minimal overhead of `PeerID` size is 160 bytes.

Also from my point of view `PeerID` is logically must be divided from peer's public key as well as from local peer's private key. Because situation looks like we trying to keep original data (e.g. private key and public key) with its hashed value.

cc @arnetheduck @zah @yglukhov @dryajov @kdeme

**Comment by arnetheduck** [comment]
I would generally agree that `PeerID` should not carry a private key:

* the majority of peers will not use the private key field (only local peer)
* we want to keep the security surface area small - that includes minimizing the places where a private key _can_ appear (in memory or otherwise)

I have no strong opinion on caching the public key - since it's only sometimes available, I don't quite see that we'll use it a lot which speaks for saving memory and reading it on the fly - I would intuitively start with a single `func extractPublicKey(pid: PeerID): Option[PublicKey]` for ease of use - if it ends up being frequently used by some particular caller, they should probably cache it - if it turns out we can motivate caching it inside `PeerID`, we can make an informed decision (based on benchmarks/real world usage) to add the cache field inside peerid later (no code changes needed for calling code)

**Comment by arnetheduck** [comment]
oh, and since publicKey is a cache right now, we run the risk of not updating it properly - another reason to start without it and add only iff motivated in real use cases

**Comment by dryajov** [comment]
I agree with the argument regarding size and efficiency. On the other hand, we're making assumptions that keys are going to be something particular and that they'd fit into the encoded multihash/multibase string. But this assumption is flawed and only holds true in certain situations, removing the public/private fields would be at best a stopgap solution, that's why I'm holding off on trying to rectify it right now.

So far, I've thought about using a `case object` that would be constructed with the correct internal structure depending on how we obtained the peer. The case object would be constructed with the private key for the local peer and hold a lazy field for resolving the public key from the private key. For incoming peers it would only expose the lazy public key field, that would either extract the encoded key from the string id or delegate to the `PublicKey` object in cases where it couldn't be found there. 

The public key is required in many places, for example Identify and PubSub and in general it is very handy to have it readily available within the peerid object.

**Comment by cheatfate** [comment]
Nimcrypto has type named `MDigest` which represent hash value, but this type do not store/cache blob of hashed data inside. `PeerID` is identical here, its just an identifier and assigning to it other roles becomes redundant. Such approach are used by Go and Rust implementations, both of them do not store Public or Private keys inside of this object.

And if you need storage or type to represent peers you can use this (you can name it as `Peer` or `PeerInfo`)
```nim
type
  PeerInfo = object
    id*: PeerID
    pubkey*: Option[PublicKey]
    addresses*: seq[MultiAddress]
```

**Comment by dryajov** [comment]
This has been addressed by #44
